### ISSUE TITLE
task.workDir is null before the within-quotes script code

### ISSUE THREAD

**Issue by Jay-uu** [nextflow-io/nextflow#4435]
## Bug report 
Hi!
Thanks for all your work making Nextflow!
I'm developing a pipeline for genome analysis and have encountered a need for being able to do local checks within a process before an if-else script block. My issue is related to #2628 and #3962, but I hope it's different enough to warrant opening a new issue.

It seems to me that the task.workDir isn't initialised until after one of the script sections start. This causes a problem when I want to interact with the input before choosing what code to run.

### Expected behavior and actual behavior

Expected behaviour: When I have a script block defined, any code within it is executed within or has access to the task.workDir.
Actual behaviour: If a conditional script block is used, any groovy code outside the quotes is executed within the launchDir, and the workDir doesn't exist yet.

### Steps to reproduce the problem
```

process mOTUs_to_pangenome {
    debug true
    input:
    path(mOTU_dir)
    shell:
    bin_list = files(task.workDir/mOTU_dir+"/*.fa") //I want this to be able to read the files from the input, but task.workDir is **null**.
    c = bin_list.size()
    single_bin = bin_list[0]
    if( c > 1 )
    '''
    #bash code
    echo "hey how come your mom lets you have two files?"
    '''
    
    else
    '''
    #!/usr/bin/env python
    #python code!
    print("ah, just one file I see")
    '''
}

 workflow {
    pg_dir = Channel.fromPath("/home/jay/c__Cyanobacteriia_mOTU_0", type: "dir", checkIfExists: true)
    mOTUs_to_pangenome(pg_dir)
}

```

### Program output 
```

ERROR ~ Error executing process > 'mOTUs_to_pangenome (1)'

Caused by:
  Cannot invoke method div() on null object -- Check script 'issue.nf' at line: 6
```

### Environment 

* Nextflow version: 23.04.3 build 5875
* Operating system: Linux
* Bash version: GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)

### Additional context

(Add any other context about the problem here)
[taskDir_null.nextflow.log](https://github.com/nextflow-io/nextflow/files/13054626/taskDir_null.nextflow.log)

**Comment by bentsherman** [comment]
I think this is because the task work dir depends on the task hash, which is not computed until after the task script is evaluated.

But you don't need to do this because `mOTU_dir` is already a path inside the task directory, so just resolve against that variable:
```groovy
bin_list = files(mOTU_dir.resolve("*.fa"))
```

**Comment by Jay-uu** [comment]
Thanks so much!
I saw some mentions about using resolve in other places but couldn't figure out the syntax. This works!

**Comment by bentsherman** [comment]
I think a better approach here would be something like this:
```groovy
process mOTUs_to_pangenome {
    debug true
    input:
    path(bin_list, arity: '1..*')
    shell:
    println("Present bins:")
    println(bin_list)
    c = bin_list.size()
    single_bin = bin_list[0]
    println("single_bin variable is:")
    println(single_bin)
    
    if( c > 1 )
    '''
    #bash code
    echo "hey how come your mom lets you have two files?"
    '''
    
    else
    '''
    #!/usr/bin/env python
    #python code!
    print("ah, just one file I see")
    '''
}

 workflow {
    pg_files = Channel.fromPath("/home/jay/c__Cyanobacteriia_mOTU_0/*.fa", checkIfExists: true)
    mOTUs_to_pangenome(pg_files)
}
```

Basically, perform the glob outside of the process. I also used the new `arity` option so that the files are always a list.

**Comment by bentsherman** [comment]
You can take this further by moving the rest of the `shell` code into channel logic in the workflow, then you could e.g. have two different processes for the different cases, but I will leave that as an exercise for the reader 😄

**Comment by Jay-uu** [comment]
Thanks for your suggestions!
I might decide to modularize the pipeline in the future and have multiple workflows - which would make the workflow check + multiple processes solution a little unstable, but it's definitely something that works now. I'll also keep a look out to see if the arity function makes it into the stable release.
Thank you for taking the time to help!

**Comment by bentsherman** [comment]
BTW the arity feature is available in 23.10.0
